compiler/rustc_parse/src/parser/path.rs

```diff
@@ -534,4 +537,51 @@ impl<'a> Parser<'a> {
         };
         Ok(Some(arg))
     }
+
+    fn get_ident_from_generic_arg(
+        &self,
+        gen_arg: Option<GenericArg>,
+        lo: Span,
+    ) -> PResult<'a, (Ident, Option<GenericArgs>)> {
+        if let Some(ref arg) = gen_arg {
+            match arg {
+                GenericArg::Type(t) => match &(*t).kind {
+                    ast::TyKind::Path(qself, ref path) => {
+                        if let Some(qself) = qself {
+                            let mut err = self.struct_span_err(
+                                qself.path_span,
+                                "invalid use of Self qualified paths",
+                            );
+                            err.span_label(
+                                qself.path_span,
+                                "not allowed in associated type constraints",
+                            );
+                            return Err(err);
+                        }
+                        if path.segments.len() == 1 {
+                            let path_seg = &path.segments[0];
+                            let ident = path_seg.ident;
+                            let gen_args = if let Some(ref gen_args) = path_seg.args {
```

Comment:
```suggestion
                            let gen_args = path_seg.args.cloned();
```
?

I think you could also try to avoid clones here, `get_ident_from_generic_arg` consumes `gen_arg` by value, so it should be potentially possible.